google/detectors/rce/CVE-2017-5638/src/main/java/com/google/tsunami/plugins/detectors/rce/CVE-2017-5638/ApacheStrutsContentTypeRceDetector.java

```diff
@@ -0,0 +1,144 @@
+  }
+
+  @Override
+  public DetectionReportList detect(
+      TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+    logger.atInfo().log(
+        "Starting Command Injection via Content-Type header (CVE-2017-5638) detection for Apache Struts.");
+    DetectionReportList detectionReports =
+        DetectionReportList.newBuilder()
+            .addAllDetectionReports(
+                matchedServices.stream()
+                    // TODO(b/147455416): checking web service is not needed once we enable
+                    // service name filtering on this plugin.
+                    .filter(NetworkServiceUtils::isWebService)
+                    .filter(this::isServiceVulnerable)
+                    .map(networkService -> buildDetectionReport(targetInfo, networkService))
+                    .collect(toImmutableList()))
+            .build();
+
+    logger.atInfo().log(
+        "ApacheStrutsContentTypeRceDetector finished, detected '%d' vulns.",
+        detectionReports.getDetectionReportsCount());
+    return detectionReports;
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
+    String targetUri = NetworkServiceUtils.buildWebApplicationRootUrl(networkService);
+
+    try {
+      // This is a blocking call.
+      String payload = String.format(PAYLOAD_STRING_FORMAT, DETECTOR_HEADER_NAME, RANDOM_VALUE);
+      HttpHeaders headers = HttpHeaders.builder()
+          .addHeader(CONTENT_TYPE, payload)
+          .build();
+      HttpResponse response = httpClient.send(get(targetUri).setHeaders(headers).build());
+      // If the server is vulnerable our header will be appended to the response.
+      Optional<String> headerValue = response.headers().get(DETECTOR_HEADER_NAME);
+
+      return headerValue.isPresent() && headerValue.get().equals(RANDOM_VALUE);
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Unable to query '%s'.", targetUri);
+      return false;
+    }
+  }
+
+  private DetectionReport buildDetectionReport(
+      TargetInfo targetInfo, NetworkService vulnerableNetworkService) {
+    return DetectionReport.newBuilder()
+        .setTargetInfo(targetInfo)
+        .setNetworkService(vulnerableNetworkService)
+        .setDetectionTimestamp(Timestamps.fromMillis(Instant.now(utcClock).toEpochMilli()))
+        .setDetectionStatus(DetectionStatus.VULNERABILITY_VERIFIED)
+        .setVulnerability(
+            Vulnerability.newBuilder()
+                .setMainId(
+                    VulnerabilityId.newBuilder()
+                        .setPublisher("GOOGLE")
+                        .setValue("CVE_2017_5638"))
+                .setSeverity(Severity.CRITICAL)
+                .setTitle("Apache Struts Command Injection via Content-Type header (CVE-2017-5638)")
+                .setDescription("Apache Struts server is vulnerable to CVE-2017-5638.")
+                .setCvssV2("10.0")
+                .setCvssV3("10.0")
```

Comment:
I think you can omit these values for now until we found a canonical way of handling CVSS scores and vectors in Tsunami.

Reply:
Done